src/logging.rs

```diff
@@ -0,0 +1,140 @@
+use std::fs::File;
+use std::io::Write;
+use std::os::unix::io::FromRawFd;
+use std::ffi::CString;
+use std::os::unix::net::UnixDatagram;
+use std::os::unix::io::{AsRawFd, IntoRawFd};
+use nix::sys::socket::{ControlMessage, MsgFlags, sendmsg};
+use nix::sys::memfd::MemFdCreateFlag;
+use nix::sys::memfd::memfd_create;
+use nix::fcntl::*;
+
+use crate::errors::*;
+
+/// Default path of the journald AF_UNIX datagram socket.
+const SD_JOURNAL_SOCK_PATH: &str = "/run/systemd/journal/socket";
+
+/// Log priority values.
+///
+/// See `man 3 syslog`.
+#[repr(u8)]
+pub enum Priority {
+    /// system is unusable
+    Emergency = 0,
+    /// action must be taken immediately
+    Alert,
+    /// critical conditions
+    Critical,
+    /// error conditions
+    Error,
+    /// warning conditions
+    Warning,
+    /// normal, but significant, condition
+    Notice,
+    /// informational message
+    Info,
+    /// debug-level message
+    Debug,
+}
+
+#[inline(always)]
+fn is_valid_char(c: char) -> bool {
+    c.is_uppercase() || c.is_numeric() || c == '_'
+}
+
+/// The variable name must be in uppercase and consist only of characters, 
+/// numbers and underscores, and may not begin with an underscore.
+fn is_valid_field(input: &str) -> bool {
+    if input.is_empty() {
+        return false;
+    }
+
+    if !input.chars().all(is_valid_char) {
+        return false;
+    }
+
+    if input.starts_with('_') {
+        return false;
+    }
+
+    true
+}
+
+fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
+    if is_valid_field(field) {
+        let field_payload = format!("{}={}\n", field, payload);
+        data.push_str(&field_payload);
+    }
+}
+
+/// Print a message to the systemd journal with the given priority.
+pub fn sd_journal_print(priority: Priority, msg: &str) -> Result<()> {
+    let sock = UnixDatagram::unbound()?;
+    sock.connect(SD_JOURNAL_SOCK_PATH)?;
+
+    let mut data = String::new();
+    add_field_and_payload(&mut data, "PRIORITY", &(priority as u8).to_string());
+    add_field_and_payload(&mut data, "MESSAGE", msg);
+
+    let res = sock.send(data.as_bytes());
+    match res {
+        Ok(_) => return Ok(()),
+        // If error code is 90, the message was too long for a UNIX socket.
+        Err(ref e) if e.raw_os_error() == Some(90) => {
+            let tmpfd = memfd_create(&CString::new("journald")?, MemFdCreateFlag::MFD_ALLOW_SEALING)?;
+            // Safe because memfd_create gave us this FD.
+            let mut file = unsafe { File::from_raw_fd(tmpfd) };
+            file.write_all(data.as_bytes())?;
+
+            fcntl(file.as_raw_fd(), FcntlArg::F_ADD_SEALS(SealFlag::all()))?;
+            let fds = &[file.into_raw_fd()];
```

Comment:
This and the line above uses the the same fd. Can you please unify these with a `let data_fd = file.into_raw_fd()` and then use `data_fd` in these two lines?